testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportResourceHungryTests.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testng.services;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.testing.ResourcePresence;
+import org.intellij.lang.annotations.Language;
+import org.testng.ITestClass;
+import org.testng.ITestContext;
+import org.testng.ITestListener;
+import org.testng.ITestNGMethod;
+import org.testng.ITestResult;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.lang.reflect.Field;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Throwables.getStackTraceAsString;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.MoreCollectors.toOptional;
+import static io.trino.testng.services.Listeners.reportListenerFailure;
+import static io.trino.testng.services.ReportResourceHungryTests.Stage.AFTER_CLASS;
+import static io.trino.testng.services.ReportResourceHungryTests.Stage.BEFORE_CLASS;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.lang.reflect.Modifier.isStatic;
+
+public class ReportResourceHungryTests
+        implements ITestListener
+{
+    private final List<Rule> rules;
+
+    public ReportResourceHungryTests()
+    {
+        ImmutableList.Builder<Rule> rules = ImmutableList.builder();
+
+        isInstance("io.trino.testing.QueryRunner").ifPresent(rules::add);
```

Comment:
This is super brittle. If these classes get moved or renamed, this code will stop doing what it should. Also, the “reverse dependency” this code now has on the rest of the system is a code smell.

Reply:
> This is super brittle. If these classes get moved or renamed, this code will stop doing what it should

of course

but it's also practical. it allowed me to find a few cases where we leak resources, something that no-one else did before 

> Also, the “reverse dependency” this code now has on the rest of the system is a code smell.

I totally agree. Do you want to propos an improvement, or just want to share that it smell bad for you too? I symphathize with that fully.

Reply:
BTW i am will see whether i can just replace this with Closeable/AutoCloseable check.
A bit of abuse of the interface (once `.close()` got called, once should not care), but that means we're on the safer side.... unless that's too many cases to fix.

Reply:
I don't have any other suggestions, but I'm concerned about the false sense of security this provides. That may go away without anyone noticing. Also, it won't work as more tests get migrated to JUnit.

Reply:
> BTW I am convinced that the benefits of the migration do not outweigh the costs, and your comment is yet another argument for that.

Which costs are you considering?